xen/arch/x86/include/asm/slaunch.h

```diff
@@ -0,0 +1,96 @@
+#ifndef _ASM_X86_SLAUNCH_H_
+#define _ASM_X86_SLAUNCH_H_
+
+#include <xen/types.h>
+#include <xen/multiboot.h>
+#include <xen/slr_table.h>
+
+#define DRTM_LOC                   2
+#define DRTM_CODE_PCR              17
+#define DRTM_DATA_PCR              18
+
+/*
+ * Secure Launch event log entry types. The TXT specification defines the
+ * base event value as 0x400 for DRTM values.
+ *
+ * Using the same values for AMD SKINIT.
+ */
+#define TXT_EVTYPE_BASE            0x400
+#define DLE_EVTYPE_SLAUNCH         (TXT_EVTYPE_BASE + 0x102)
+#define DLE_EVTYPE_SLAUNCH_START   (TXT_EVTYPE_BASE + 0x103)
+#define DLE_EVTYPE_SLAUNCH_END     (TXT_EVTYPE_BASE + 0x104)
+
+#ifndef cpuid
+/*
+ * Generic CPUID function
+ * clear %ecx since some cpus (Cyrix MII) do not set or clear %ecx
+ * resulting in stale register contents being returned.
+ *
+ * Copied from processor.h because that header can't be included by early code.
+ */
+#define cpuid(_op,_eax,_ebx,_ecx,_edx)          \
+    asm volatile ( "cpuid"                      \
+          : "=a" (*(int *)(_eax)),              \
+            "=b" (*(int *)(_ebx)),              \
+            "=c" (*(int *)(_ecx)),              \
+            "=d" (*(int *)(_edx))               \
```

Comment:
This is out of date.  https://xenbits.xen.org/gitweb/?p=xen.git;a=commitdiff;h=86d0edda34e732f1b000d2adddf4f7884d18e04a finally purged the last of the macros for MISRA reasons.

In order to get code accepted into upstream Xen, it's going to need to be rebased onto the `staging` branch, and rearranged such that it can be integrated cleanly.

I'm happy to help out with whatever prep is necessary, and to provide some recommendations for how to structure the rest of the series.

Do you have a single branch with all work on it, rather than a delta against a still-pending delta?

Reply:
@krystian-hebel There are two such uses and each of them is invoked once resulting in 16 memory accesses in total, is that a lot?  If so, you might want to look at `objdump -d xen/arch/x86/boot/slaunch_early.o`, for example:
```asm
 615:   89 45 e8                mov    %eax,-0x18(%ebp)
 618:   8b 45 f4                mov    -0xc(%ebp),%eax
 61b:   89 44 24 10             mov    %eax,0x10(%esp)
 61f:   8b 45 0c                mov    0xc(%ebp),%eax
 622:   89 44 24 0c             mov    %eax,0xc(%esp)
 626:   8b 45 08                mov    0x8(%ebp),%eax
 629:   89 44 24 08             mov    %eax,0x8(%esp)
 62d:   8b 45 e8                mov    -0x18(%ebp),%eax
 630:   89 44 24 04             mov    %eax,0x4(%esp)
 634:   8b 45 ec                mov    -0x14(%ebp),%eax
 637:   89 04 24                mov    %eax,(%esp)
```
Early C code seems to be compiled without optimizations.

Reply:
There is also implicit read done by `cmp` instructions. This is infinity times more than what is really necessary :slightly_smiling_face: In any case, as Andrew mentioned, we (will) have to abide to MISRA rules, so let's leave it as it is.

> In order to get code accepted into upstream Xen, it's going to need to be rebased onto the `staging` branch, and rearranged such that it can be integrated cleanly.

> Do you have a single branch with all work on it, rather than a delta against a still-pending delta?

We do not. Right now this is based on whatever Qubes OS uses, because this is what we're obligated to as part of [this project](https://docs.dasharo.com/projects/trenchboot-aem-v2/), any potential upstreaming is extra (as in: no deadline so lower priority).